core/modules/replica/Job.cc

```diff
@@ -0,0 +1,435 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/Job.h"
+
+// System headers
+#include <sstream>
+#include <stdexcept>
+#include <utility>      // std::swap
+
+// Third party headers
+#include <boost/bind.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+
+// Qserv headers
+#include "lsst/log/Log.h"
+#include "replica/AddReplicaQservMgtRequest.h"
+#include "replica/Common.h"            // Generators::uniqueId()
+#include "replica/Configuration.h"
+#include "replica/DatabaseServices.h"
+#include "replica/Performance.h"       // PerformanceUtils::now()
+#include "replica/QservMgtServices.h"
+#include "replica/RemoveReplicaQservMgtRequest.h"
+#include "replica/ServiceProvider.h"
+#include "util/IterableFormatter.h"
+
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.Job");
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+std::string Job::state2string(State state) {
+    switch (state) {
+        case CREATED:     return "CREATED";
+        case IN_PROGRESS: return "IN_PROGRESS";
+        case FINISHED:    return "FINISHED";
+    }
+    throw std::logic_error(
+                "incomplete implementation of method Job::state2string(State)");
+}
+
+std::string
+Job::state2string(ExtendedState state) {
+    switch (state) {
+        case NONE:         return "NONE";
+        case SUCCESS:      return "SUCCESS";
+        case CONFIG_ERROR: return "CONFIG_ERROR";
+        case FAILED:       return "FAILED";
+        case QSERV_FAILED: return "QSERV_FAILED";
+        case QSERV_IN_USE: return "QSERV_IN_USE";
+        case EXPIRED:      return "EXPIRED";
+        case CANCELLED:    return "CANCELLED";
+    }
+    throw std::logic_error(
+                "incomplete implementation of method Job::state2string(ExtendedState)");
+}
+
+Job::Job(Controller::Ptr const& controller,
+         std::string const& parentJobId,
+         std::string const& type,
+         Options const& options)
+    :   _id(Generators::uniqueId()),
+        _controller(controller),
+        _parentJobId(parentJobId),
+        _type(type),
+        _options(options),
+        _state(State::CREATED),
+        _extendedState(ExtendedState::NONE),
+        _beginTime(0),
+        _endTime(0),
+        _heartbeatTimerIvalSec(controller->serviceProvider()->config()->jobHeartbeatTimeoutSec()),
+        _heartbeatTimerPtr(nullptr),
+        _expirationIvalSec(controller->serviceProvider()->config()->jobTimeoutSec()),
+        _expirationTimerPtr(nullptr) {
+}
+
+Job::Options Job::setOptions(Options const& newOptions) {
+    Options options(newOptions);
+    std::swap(_options, options);
```

Comment:
I don't think this is thread safe; another user could be using `_options` (returned by reference from `options()`) and this would cause _options to be deleted.

It looks like this is never called. Are you sure it's necessary?

Reply:
reinforced method options() to be thread-safe

Method setOptions() is not presently used but it will be used in the future. So, I'd leave it for now.